module/mempool/queue/messageEntity.go

```diff
@@ -0,0 +1,52 @@
+package queue
+
+import (
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"strconv"
+	"time"
+)
+
+// MessageEntity is a data structure for storing messages in HeroQueue.
+type MessageEntity struct {
+	Msg engine.Message
+	id  flow.Identifier
+}
+
+var _ flow.Entity = (*MessageEntity)(nil)
+
+// NewMessageEntity returns a new message entity.
+func NewMessageEntity(msg *engine.Message) MessageEntity {
+	id := identifierOfMessage(msg)
+	return MessageEntity{
+		Msg: *msg,
+		id:  id,
+	}
+}
+
+// NewMessageEntityWithNonce creates a new message entity adding a nonce to the id calculation.
```

Comment:
```suggestion
// NewMessageEntityWithNonce creates a new message entity adding a nonce to the id calculation.
// This prevents de-duplication of otherwise identical messages stored in the queue. 
```